### PR TITLE
Fix #3: Language parser

### LINKED ISSUE #3: Improve message parsing
The current parsing flow is very rigid. It should allow for more free statements like:
- @osdc-bot joke
- @osdc-bot tell me a joke
- @osdc-bot joke suna
- @osdc-bot can i hear a joke

Parsing should be done such that all the above should yield a joke.

### PR DESCRIPTION
**Pull request by wimpykid26**
Fixes #3 
@526avijitgupta Please review.

UPDATE:
Tasklist:
- [x] Move api calls to js and remove python code.
- [ ] ~~Integrate nltk with server.py.~~

### CODE REVIEW

File: services/autocomplete.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import urllib.request
+import json
+def message3(var):
```
Comment (526avijitgupta) [inline services/autocomplete.py]: @wimpykid26  I would highly suggest removing this file altogether and re-writing this code in JS. Create a new file similar to https://github.com/osdc/osdc-bot/blob/master/services/jokeService.js (in the directory services/) and add another if condition in chatbot.js (see https://github.com/osdc/osdc-bot/blob/master/chatbot.js#L105-L107) to simply call that method from the 1st file if the string matching succeeds!

File: services/crow.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import urllib.request
+import json
+def message(var):
```
Comment (526avijitgupta) [inline services/crow.py]: See comment above.

File: services/joke.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import urllib.request
+import json
+def message2():
```
Comment (526avijitgupta) [inline services/joke.py]: Same.

**Comment by yashmehrotra** [comment]
@526avijitgupta It can't be done in JS, nltk is for python only

**Comment by 526avijitgupta** [comment]
@wimpykid26 Left a few initial review comments. Please resolve them and mark for review again when that's done.
Also, feel free to ping me or @yashmehrotra here or on Gitter if you need help with this 😄

**Comment by 526avijitgupta** [comment]
@yashmehrotra Only 1 out of the 4 files is using nltk. The rest are simply calling different api's and returning the required data -- that can be done in js.

**Comment by yashmehrotra** [comment]
@wimpykid26 
1. Make the directory structure like, services/insult_joke_service/..., services/xyz_service/...
2. Make your code compatible to: https://www.python.org/dev/peps/pep-0008/.
3. Please name the functions and variables in a better way

**Comment by yashmehrotra** [comment]
@wimpykid26 If possible, write the API code JS, and the NLP logic in server.py for now,
they'll communicate through HTTP

**Comment by wimpykid26** [comment]
@yashmehrotra @526avijitgupta Please review.
^Updated checklist in intial comment as well.

File: chatbot.js [hunk 4]
```diff
@@ -2,7 +2,10 @@ const Faye = require('faye');
 const request = require('request');
 
 const config = require('./config');
-const joker = require('./services/jokeService.js');
+const joker = require('./services/jokeService');
+const weather = require('./services/weather');
```
Comment (526avijitgupta) [inline chatbot.js]: rename to weatherService for consistency.

File: chatbot.js [hunk 5]
```diff
@@ -22,7 +25,10 @@ const BOT_ACTIONS = {
   HELP: 'help',
   JOKE: 'joke',
   DEPLOY: 'deploy',
-  HOWDOI: 'howdoi'
+  HOWDOI: 'howdoi',
+  WEATHER: 'weather',
+  PLACES: 'locate'
+
```
Comment (526avijitgupta) [inline chatbot.js]: Remove newline.

File: chatbot.js [hunk 6]
```diff
@@ -2,7 +2,10 @@ const Faye = require('faye');
 const request = require('request');
 
 const config = require('./config');
-const joker = require('./services/jokeService.js');
+const joker = require('./services/jokeService');
+const weather = require('./services/weather');
+const places = require('./services/mapService');
+
```
Comment (526avijitgupta) [inline chatbot.js:9]: Remove newline.

File: chatbot.js [hunk 7]
```diff
@@ -93,11 +98,11 @@ function _getBotHelp() {
...
   const displayName = user.displayName;
   const username = user.username;
-
   if (message.startsWith(BOT_MENTION_NAME)) {
     const parsedMessage = message.slice(BOT_MENTION_NAME.length + 1);
-    console.log(parsedMessage);
     const startsWithString = _getStartsWith(parsedMessage);
+    const splitMessage = parsedMessage.split(' ').splice(1);
+    const cityName = `${splitMessage.join(' ')}`;
     if (startsWithString === BOT_ACTIONS.HELP) {
```
Comment (526avijitgupta) [inline chatbot.js]: replace by
```const cityName = splitMessage.join(' ');```

File: chatbot.js [hunk 8]
```diff
@@ -93,11 +98,11 @@ function _getBotHelp() {
 function reply_to_user(user, message) {
   const displayName = user.displayName;
   const username = user.username;
-
   if (message.startsWith(BOT_MENTION_NAME)) {
     const parsedMessage = message.slice(BOT_MENTION_NAME.length + 1);
-    console.log(parsedMessage);
     const startsWithString = _getStartsWith(parsedMessage);
+    const splitMessage = parsedMessage.split(' ').splice(1);
+    const cityName = `${splitMessage.join(' ')}`;
```
Comment (526avijitgupta) [inline chatbot.js]: replace by
`const cityName = splitMessage.join(' ');`

Comment (526avijitgupta) [reply chatbot.js]: Even better -
You don't need the variable `splitMessage` at all. Replace both lines by

``` js
const cityName = parsedMessage.substr(parsedMessage.indexOf(' ') + 1);
```

**Comment by 526avijitgupta** [comment]
@wimpykid26 Took a review pass. Please try and address the review comments before proceeding further!

**Comment by wimpykid26** [comment]
@526avijitgupta please review the latest commit

File: chatbot.js [hunk 9]
```diff
@@ -93,11 +96,10 @@ function _getBotHelp() {
 function reply_to_user(user, message) {
   const displayName = user.displayName;
   const username = user.username;
-
   if (message.startsWith(BOT_MENTION_NAME)) {
     const parsedMessage = message.slice(BOT_MENTION_NAME.length + 1);
-    console.log(parsedMessage);
     const startsWithString = _getStartsWith(parsedMessage);
+    const cityName = parsedMessage.substr(parsedMessage.indexOf(' ')+1);
     if (startsWithString === BOT_ACTIONS.HELP) {
```
Comment (526avijitgupta) [inline chatbot.js]: spaces around +

File: chatbot.js [hunk 10]
```diff
@@ -93,11 +96,10 @@ function _getBotHelp() {
 function reply_to_user(user, message) {
   const displayName = user.displayName;
   const username = user.username;
-
   if (message.startsWith(BOT_MENTION_NAME)) {
     const parsedMessage = message.slice(BOT_MENTION_NAME.length + 1);
-    console.log(parsedMessage);
     const startsWithString = _getStartsWith(parsedMessage);
+    const cityName = parsedMessage.substr(parsedMessage.indexOf(' ')+1);
```
Comment (526avijitgupta) [inline chatbot.js:105]: spaces around +

**Comment by 526avijitgupta** [comment]
@wimpykid26 Great work. Task 1 seems to be completed (just a minor comment)!

**Comment by 526avijitgupta** [comment]
@yashmehrotra As discussed with @wimpykid26 -- Task 2 will be done in a separate PR. He'll remove the .py file for nltk for now. I guess this should be mergeable once the py file is removed and the line comment is addressed. wdyt ?

**Comment by yashmehrotra** [comment]
@526avijitgupta Yeah, you are right.
@wimpykid26 Things to do:
1. Address the line comments made by Avijit.
2. Remove the .py file
3. Squash everything into a single commit

Bonus: Were you able to add Google Map link with the location command ?

**Comment by 526avijitgupta** [comment]
@wimpykid26 You may skip the 3rd point as mentioned by Yash, if you haven't done that already.
/cc @yashmehrotra - GitHub offers a squash and merge option!

File: services/weatherService.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
+const request = require('request');
+
+module.exports = {
+  getWeather: (callback, username, city_name) => {
+    const encodedCityName = encodeURIComponent(city_name);
```
Comment (526avijitgupta) [inline services/weatherService.js]: Replace `city_name` by `cityName` on both lines.

**Comment by wimpykid26** [comment]
@526avijitgupta Removed the python script. Please review the latest commit

File: services/weatherService.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    const WEATHER_API_URL = `http://api.openweathermap.org/data/2.5/weather?q=${encodedCityName}&APPID=5ce8ec77d11e6b31bbca4a128afd3b6d`;
+    
+    request(WEATHER_API_URL, (error, response, body) => {
+      if (!error && response.statusCode == 200) {
+        const weatherData = JSON.parse(body);
+        const KELVIN_CELCIUS_OFFSET = 273;
+        const maxTempCelcius = (weatherData.main.temp_max - KELVIN_CELCIUS_OFFSET).toFixed(2);
+        const minTempCelcius = (weatherData.main.temp_min - KELVIN_CELCIUS_OFFSET).toFixed(2);
+        message = `Humidity: ${weatherData.main.humidity}%\nMax Temperature: ${maxTempCelcius}C\nMin Temperature: ${minTempCelcius}C`;
```
Comment (526avijitgupta) [inline services/weatherService.js]: `const message = ...`

File: services/weatherService.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
+const request = require('request');
+
+module.exports = {
+  getWeather: (callback, username, city_name) => {
+    const encodedCityName = encodeURIComponent(city_name);
+    const WEATHER_API_URL = `http://api.openweathermap.org/data/2.5/weather?q=${encodedCityName}&APPID=5ce8ec77d11e6b31bbca4a128afd3b6d`;
+    
+    request(WEATHER_API_URL, (error, response, body) => {
+      if (!error && response.statusCode == 200) {
```
Comment (526avijitgupta) [inline services/weatherService.js]: use `===`

File: chatbot.js [hunk 14]
```diff
@@ -81,7 +85,6 @@ function _getStartsWith(parsedMessage) {
   }
   return result;
 }
-
 function _getBotHelp() {
```
Comment (526avijitgupta) [inline chatbot.js:91]: Leave newline above this function.

**Comment by wimpykid26** [comment]
@526avijitgupta please review the latest commit

File: chatbot.js (hunk 10 above)
Comment (526avijitgupta) [reply chatbot.js:105]: @wimpykid26 I guess you missed this one.

**Comment by 526avijitgupta** [comment]
@wimpykid26 See https://github.com/osdc/osdc-bot/pull/4#discussion_r77445180 .

**Comment by wimpykid26** [comment]
@526avijitgupta I cant seem to find the spaces in line 102.

**Comment by deathping1994** [comment]
how about using something simple like naive bayesian classifier. Why bring out the big guns like nltk for a simple task ?

**Comment by yashmehrotra** [comment]
@deathping1994 @wimpykid26 is taking point on this one, you can discuss it directly with him on our gitter channel

**Pull request merged**